src/Npgsql/NpgsqlParameterCollection.cs

```diff
@@ -570,5 +673,48 @@ static NpgsqlParameter Cast(object? value)
                 throw new InvalidCastException($"The value \"{value}\" is not of type \"{nameof(NpgsqlParameter)}\" and cannot be used in this parameter collection.");
             }
         }
+
+        struct MultiValue
+        {
+            public int Value { get; set; }
+            public List<int>? Values { get; private set; }
+
+            [MemberNotNullWhen(true, nameof(Values))]
+            public bool HasMultiple => Values is { Count: > 1 };
+
+            public MultiValue(int value)
+            {
+                Value = value;
+                Values = null;
+            }
+
+            public void Add(int value)
+            {
+                if (HasMultiple)
+                {
+                    Values.Add(value);
+                    return;
+                }
+                Values ??= new List<int>(2);
+                Values.Add(Value);
+                Values.Add(value);
+                Value = default;
+            }
+
+            public MultiValue? Remove(int index)
+            {
+                if (!HasMultiple)
+                    return null;
+
+                // Demote the MultiValue but keep the list allocated.
+                if (Values.Remove(index) && !HasMultiple)
```

Comment:
I tried to keep allocations amortized instead of optimizing for fastest lookup perf. If you hit these paths you're probably doing a lot of mutations on your param collection, better to keep things steady then.